### DISCUSSION TITLE
Use CI_JOB_TOKEN for access in Gitlab using new token access feature in 15.9

### DISCUSSION THREAD

**Discussion by roemba** [renovatebot/renovate#20336] [Suggest an Idea]
### Type of discussion.

I'm proposing an idea

### Tell us more.

The [Renovate documentation](https://docs.renovatebot.com/gitlab-bot-security/#status-of-the-renovate-app-for-gitlab) mentions a [Gitlab issue](https://gitlab.com/gitlab-org/gitlab/-/issues/346298) that has now been closed and the feature it mentions is deployed in 15.9

It seems a relatively small change to make in Renovate where it will use the CI_JOB_TOKEN for API access if so configured. The CI_JOB_TOKEN for the project it is running in can then be used to access other projects by configuring some settings in Gitlab. See https://docs.gitlab.com/ee/ci/jobs/ci_job_token.html#add-a-project-to-the-inbound-job-token-scope-allowlist

This will allow running Renovate in Gitlab without creating a user for it which can be very useful if a user pays a license with limited seats.

**Comment by rarkins** [comment]
I'm not sure what you mean specifically. Renovate doesn't force which type of token you use today.

**Reply by viceice** [reply]
@rarkins we should configure this for our runner template to make the workflow easier

**Reply by roemba** [reply]
I know the type of token is not forced, but I was under the impression that Renovate (when configured with `platform = gitlab`) uses the User API to get access to other repositories, which won't work with a CI_JOB_TOKEN. But I'll be the first to admit that I don't know Renovate well enough to know if there is a also a configuration option for that.

**Reply by viceice** [reply]
https://gitlab.com/renovate-bot/renovate-runner/-/merge_requests/202

that are my tries to get it working with a job token. 😕

**Reply by Vringe** [reply]
Renovate tries to fetch user info which fails because the CI_JOB_TOKEN has no permission to do that.

```
DEBUG: GET https://gitlab.internal/api/v4/user = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=401 retryCount=0, duration=105)
```

**Reply by viceice** [reply]
I've configured the `username` in my tests and renovate didn't called that api but failed at a different one.

**Reply by Vringe** [reply]
Interesting. I just tried setting `RENOVATE_USERNAME: $GITLAB_USER_LOGIN` but it still tries to fetch `/api/v4/user`

**Reply by viceice** [reply]
try to set `gitAuthor` too

https://github.com/renovatebot/renovate/blob/4aa94080411f81c4a3aed319640d9ccca8c9299c/lib/modules/platform/gitlab/index.ts#L128-L141

**Reply by Vringe** [reply]
Thank you for the hint. I managed to get alot further. Now Renovate wants to fetch the repository via `/api/v4/projects/group/my-project` which fails with a 404 because the CI_JOB_TOKEN [does not have permission](https://docs.gitlab.com/ci/jobs/ci_job_token/) to access this endpoint. Did you find a way to get through this or am I at the same point you were stuck?

This is what I currently have:
```
    RENOVATE_PLATFORM: gitlab
    RENOVATE_ENDPOINT: $CI_API_V4_URL
    RENOVATE_TOKEN: $CI_JOB_TOKEN
    RENOVATE_USERNAME: $GITLAB_USER_LOGIN
    RENOVATE_GIT_AUTHOR: "Renovate Bot <$GITLAB_USER_EMAIL>"
    RENOVATE_X_PLATFORM_VERSION: $CI_SERVER_VERSION
    RENOVATE_REPOSITORIES: '["$CI_PROJECT_PATH"]'
    RENOVATE_DEFAULT_BRANCH: main
```

**Reply by viceice** [reply]
probably same point 😞

**Reply by Vringe** [reply]
It's a shame that you can't make renovate just use the already checked out repository inside the pipeline. That would make it alot easier. I know there is the [local](https://docs.renovatebot.com/modules/platform/local/) platform but it's experimental and does not look like it's designed for that.

**Reply by imdmahajankanika** [reply]
Hello! I also got 404 Repository is not found

 ```
INFO: Renovate started
       "renovateVersion": "43.256.1"
 INFO: Repository started (repository=xxx/xxx-2022/back)
       "renovateVersion": "43.256.1"
ERROR: Repository is not found (repository=xxx/xxx-2022/back)
 INFO: Repository finished (repository=xxx/xxx-2022/back)
       "cloned": undefined,
       "durationMs": 213,
       "result": "not-found",
       "status": "disabled",
       "enabled": false,
       "onboarded": undefined
 INFO: Renovate was run at log level "info". Set LOG_LEVEL=debug in environment variables to see extended debug logs.
 INFO: Renovate is exiting with a non-zero code due to the following logged errors
       "loggerErrors": [
         {
           "name": "renovate",
           "level": 50,
           "logContext": "c741f3de-9e4e-48a4-ab5d-97504261db49",
           "repository": "xxx/xxx-2022/back",
           "msg": "Repository is not found"
         }
       ]
```

**Reply by RahulGautamSingh** [reply]
@imdmahajankanika Please create a separate discussion with job logs.
